lib/hmmlearn/kl_divergence.py

```diff
@@ -0,0 +1,147 @@
+"""
+
+All implementations are based upon the following:
+    http://www.fil.ion.ucl.ac.uk/~wpenny/publications/densities.ps
+"""
+
+import numpy as np
+
+from scipy.special import gammaln, digamma
+from sklearn.utils import check_array
+
+
+def kl_dirichlet(q, p):
+    """
+    KL Divergence between two dirichlet distributions
+
+    KL(q || p) = ln [gamma(q)/gamma(p)] - sum [ ln [gamma(q_j)/gamma(p_j)]
+                 - (q_j - p_j) (digamma(q_j) - digamma(p_j)]
+
+    """
+    q = np.asarray(q)
+    p = np.asarray(p)
+
+    qsum = q.sum()
+    psum = p.sum()
+    delta = q - p
+    return gammaln(qsum) - gammaln(psum) - \
+            np.sum(gammaln(q) - gammaln(p)) + \
+            np.sum(
+                np.dot(
```

Comment:
why is there both a sum and a dot?  are q and p intended to be 2D?

Reply:
I'll take a few days to go over these and ping you.

Reply:
Cleaned up using einsum.